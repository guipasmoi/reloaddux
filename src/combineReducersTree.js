/* eslint no-param-reassign: "off" */
import isArray from "lodash/isArray";
import isPlainObject from "lodash/isPlainObject";
import set from "lodash/set";
import merge from "lodash/merge";

function isValidNode(node) {
  return isPlainObject(node);
}

function isActionsArray(actions) {
  return actions === undefined ||
    (isArray(actions) &&
      actions.reduce(
        (accu, action) => accu && typeof action === "string",
        true
      ));
}

function isLeaf(node) {
  return node &&
    typeof node === "object" &&
    isActionsArray(node.actions) &&
    typeof node.reducer === "function";
}

export default function combineReducersTree(
  tree,
  initAction = { type: "@@redux/INIT" }
) {
  const reversedTree = {
    // action1: { scope1: reducer, scope2: reducer },
    "*": {}
  };

  let hasBeenInitialized = false;

  function reverseTree(subTree, scope = "") {
    Object.entries(subTree).forEach(([key, node]) => {
      if (isLeaf(subTree)) {
        if (subTree.actions) {
          subTree.actions.forEach(action => {
            if (reversedTree[action] === undefined) {
              reversedTree[action] = {};
            }
            set(reversedTree[action], scope, subTree);
          });
        } else {
          set(reversedTree["*"], scope, subTree);
        }
      } else if (isValidNode(subTree)) {
        reverseTree(node, scope.length === 0 ? key : `${scope}.${key}`);
      }
    });
  }

  reverseTree(tree);
  // return reversedTree;
  function recursiveProcess(state, action, task) {
    if (isLeaf(task)) {
      const newState = task.reducer(state, action);
      return { value: newState, hasChanged: newState !== state };
    } else if (!isPlainObject(task)) {
      // reversedTree (task) must contain only plain object and reducer
      throw new Error("unexpected");
    }
    const resultValue = Object.entries(task).reduce(
      (accu, [key, subtask]) => {
        const temp = recursiveProcess(
          isPlainObject(state) ? state[key] : {},
          action,
          subtask
        );
        accu.value[key] = temp.value;
        accu.hasChanged = accu.hasChanged || temp.hasChanged;
        return accu;
      },
      { value: { ...state }, hasChanged: false }
    );
    return {
      ...resultValue,
      hasChanged: resultValue.hasChanged || resultValue.value !== state
    };
  }

  // todo optmization no avoid merge on each action
  return (state, action) => {
    let temp = { value: state, hasChanged: false };
    if (!hasBeenInitialized || initAction.type === action.type) {
      temp = recursiveProcess(state, initAction, tree);
      hasBeenInitialized = true;
    }
    temp = recursiveProcess(
      temp.value,
      action,
      merge({}, reversedTree[action.type], reversedTree["*"])
    );
    return temp.hasChanged ? temp.value : state;
  };
}
