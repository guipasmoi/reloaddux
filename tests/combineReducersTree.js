import combineReducersTree from "../src/combineReducersTree";

const data = {
  groceryManagement: {
    data: {
      groceries: {
        actions: ["ACTION1"],
        reducer: (state, action) => action.data
      }
    },
    container: {
      currentList: {
        actions: ["ACTION3", "ACTION4"],
        reducer: (state, action) => action.data
      }
    }
  },
  listManagement: {
    data: {
      lists: {
        reducer: (state, action) => action.data
      }
    },
    container: {
      groceryListsEdit: {
        actions: ["ACTION6"],
        reducer: (state, action) => action.data
      }
    }
  },
  common: {
    container: {
      application: {
        actions: ["ACTION7"],
        reducer: (state, action) => action.data
      },
      collapsableSidebar: {
        reducer: (state, action) => action.data
      }
    }
  }
};

const initialState = {
  groceryManagement: {
    data: {
      groceries: "some data at start"
    }
  },
  basket: "a basket that shouldn't be touched",
  building: {
    appartment1: "it s my home must not touch it"
  }
};

describe("combineReducerTree", () => {
  it("It can init state with init action", () => {
    const a = combineReducersTree(data);
    expect(a({}, { type: "@@redux/INIT" })).toMatchSnapshot();
  });

  it("It can dispatch a scpecific action", () => {
    const a = combineReducersTree(data);
    expect(
      a({}, { type: "ACTION1", data: "some data after ACTION1" })
    ).toMatchSnapshot();
  });

  it("It can dispatch a scpecific action with a previous state", () => {
    const a = combineReducersTree(data);
    expect(
      a(initialState, { type: "ACTION1", data: "some data after ACTION1" })
    ).toMatchSnapshot();
  });

  it("Previous state is not alterated by action and immutability principle are respected", () => {
    const a = combineReducersTree(data);
    const initialStateBefore = JSON.stringify(initialState);
    const newState = a(initialState, {
      type: "ACTION1",
      data: "some data after ACTION1"
    });
    const initialStateAfter = JSON.stringify(initialState);
    expect(initialStateBefore).toBe(initialStateAfter);
    expect(initialState).not.toBe(newState);
    expect(initialState.building.appartment1).toBe(
      newState.building.appartment1
    );
    expect(initialState.building.appartment1).toBe(
      newState.building.appartment1
    );
    expect(newState).toMatchSnapshot();
    expect(initialState.groceryManagement.data.groceries).not.toBe(
      newState.groceryManagement.data.groceries
    );
    expect(initialState.groceryManagement).not.toBe(newState.groceryManagement);
    expect(initialState.basket).toBe(newState.basket);
  });
});
