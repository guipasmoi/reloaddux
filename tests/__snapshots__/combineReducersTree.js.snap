// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combineReducerTree It can dispatch a scpecific action 1`] = `
Object {
  "common": Object {
    "container": Object {
      "application": undefined,
      "collapsableSidebar": "some data after ACTION1",
    },
  },
  "groceryManagement": Object {
    "container": Object {
      "currentList": undefined,
    },
    "data": Object {
      "groceries": "some data after ACTION1",
    },
  },
  "listManagement": Object {
    "container": Object {
      "groceryListsEdit": undefined,
    },
    "data": Object {
      "lists": "some data after ACTION1",
    },
  },
}
`;

exports[`combineReducerTree It can dispatch a scpecific action with a previous state 1`] = `
Object {
  "basket": "a basket that shouldn't be touched",
  "building": Object {
    "appartment1": "it s my home must not touch it",
  },
  "common": Object {
    "container": Object {
      "application": undefined,
      "collapsableSidebar": "some data after ACTION1",
    },
  },
  "groceryManagement": Object {
    "container": Object {
      "currentList": undefined,
    },
    "data": Object {
      "groceries": "some data after ACTION1",
    },
  },
  "listManagement": Object {
    "container": Object {
      "groceryListsEdit": undefined,
    },
    "data": Object {
      "lists": "some data after ACTION1",
    },
  },
}
`;

exports[`combineReducerTree It can init state with init action 1`] = `
Object {
  "common": Object {
    "container": Object {
      "application": undefined,
      "collapsableSidebar": undefined,
    },
  },
  "groceryManagement": Object {
    "container": Object {
      "currentList": undefined,
    },
    "data": Object {
      "groceries": undefined,
    },
  },
  "listManagement": Object {
    "container": Object {
      "groceryListsEdit": undefined,
    },
    "data": Object {
      "lists": undefined,
    },
  },
}
`;

exports[`combineReducerTree Previous state is not alterated by action and immutability principle are respected 1`] = `
Object {
  "basket": "a basket that shouldn't be touched",
  "building": Object {
    "appartment1": "it s my home must not touch it",
  },
  "common": Object {
    "container": Object {
      "application": undefined,
      "collapsableSidebar": "some data after ACTION1",
    },
  },
  "groceryManagement": Object {
    "container": Object {
      "currentList": undefined,
    },
    "data": Object {
      "groceries": "some data after ACTION1",
    },
  },
  "listManagement": Object {
    "container": Object {
      "groceryListsEdit": undefined,
    },
    "data": Object {
      "lists": "some data after ACTION1",
    },
  },
}
`;
